NIRSpec FS Pipeline Demo Notebook

kglidic opened the pull request:
This notebook checklist has been made available to us by the [the Notebooks For All team](https://github.com/Iota-School/notebooks-for-all/blob/main/resources/event-hackathon/notebook-authoring-checklist.md).
Its purpose is to serve as a guide for both the notebook author and the technical reviewer highlighting critical aspects to consider when striving to develop an accessible and effective notebook.

### The First Cell

- [ ] The title of the notebook in a first-level heading (eg. `<h1>` or `# in markdown`).
- [ ] A brief description of the notebook.
- [ ] A table of contents in an [ordered list](https://www.markdownguide.org/basic-syntax/#ordered-lists) (`1., 2.,` etc. in Markdown).
- [ ] The author(s) and affiliation(s) (if relevant).
- [ ] The date first published.
- [ ] The date last edited (if relevant).
- [ ] A link to the notebook's source(s) (if relevant).

### The Rest of the Cells

- [ ] There is only one H1 (`#` in Markdown) used in the notebook.
- [ ] The notebook uses other [heading tags](https://www.markdownguide.org/basic-syntax/#headings) in order (meaning it does not skip numbers). 

## Text

- [ ] All link text is descriptive. It tells users where they will be taken if they open the link.
- [ ] All acronyms are defined at least the first time they are used. 
- [ ] Field-specific/specialized terms are used when needed, but not excessively.

## Code

- [ ] Code sections are introduced and explained before they appear in the notebook. This can be fulfilled with a heading in a prior Markdown cell, a sentence preceding it, or a code comment in the code section.
- [ ] Code has explanatory comments (if relevant). This is most important for long sections of code.
- [ ] If the author has control over the syntax highlighting theme in the notebook, that theme has enough color contrast to be legible.
- [ ] Code and code explanations focus on one task at a time. Unless comparison is the point of the notebook, only one method for completing the task is described at a time.

### Images

- [ ] All images (jpg, png, svgs) have an [image description](https://www.w3.org/WAI/tutorials/images/decision-tree/). This could be
    - [ ] [Alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt) (an `alt` property) 
    - [ ] [Empty alt text](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/alt#decorative_images) for decorative images/images meant to be skipped (an `alt` attribute with no value)
    - [ ] Captions
    - [ ] If no other options will work, the image is decribed in surrounding paragraphs.

- [ ] Any [text present in images](https://www.w3.org/WAI/WCAG21/Understanding/images-of-text.html) exists in a text form outside of the image (this can be alt text, captions, or surrounding text.)

### Visualizations

- [ ] All visualizations have an image description. Review the previous section, Images, for more information on how to add it.
- [ ] [Visualization descriptions](http://diagramcenter.org/specific-guidelines-e.html) include
    - [ ] The type of visualization (like bar chart, scatter plot, etc.)
    - [ ] Title
    - [ ] Axis labels and range
    - [ ] Key or legend
    - [ ] An explanation of the visualization's significance to the notebook (like the trend, an outlier in the data, what the author learned from it, etc.)

- [ ] All visualizations and their parts have [enough color contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) ([color contrast checker](https://webaim.org/resources/contrastchecker/)) to be legible. Remember that transparent colors have lower contrast than their opaque versions.
- [ ] All visualizations [convey information with more visual cues than color coding](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html). Use text labels, patterns, or icons alongside color to achieve this.
- [ ] All visualizations have an additional way for notebook readers to access the information. Linking to the original data, including a table of the data in the same notebook, or sonifying the plot are all options.

drlaw1558 commented:
Tagging @rizeladiaz

kglidic commented:
I have updated the notebook to use the latest pipeline version 1.16.0 and correct CRDS context.

rizeladiaz commented:
Started reviewing this notebook

rizeladiaz commented:
I completed the review of this notebook. Just waiting for @kglidic 's approval.

rizeladiaz commented on notebooks/NIRSPEC/FS_demo.ipynb:
The notebook and requirements file here have not been updated. The changes are waiting to be merged in kglidic branch, after which, she should submit a new pull request to spacetelescope.

kglidic replied:
Sorry for the delay while I was out of office. Pull request to spacetelescope has been updated with the changes. I made a comment detailing any additional changes I made based on review.

kglidic commented:
Thank you, @rizeladiaz, for thoroughly reviewing this notebook. I’ve incorporated your feedback and made a few changes:

1.  I addressed formatting and spelling issues and removed repetitive timestamps.
2. Previously, the notebook contained some code that compared the results to those of MAST products. I’ve removed any remaining calls to this (but it remains a parameter in the helper function). 
3.  I added print statements to compare the default CRDS context associated with the build with the context actively set in the notebook as a check.
4. Examples for Creating Association (ASN) Files: I added a helper function, write_asn, with examples for manually creating ASN files for spec2 and spec3. This function was part of an earlier version of the notebook I worked on, but I removed it after discussions about how some observations might have more complex ASN files (which my current function may not be able to handle). For example, the current function can handle spectral dithers and dedicated background exposures when creating spec2 ASN files. But, it would likely not handle spatial dithers properly. So, by default, the notebook is set to download the ASN files from MAST, but if the user chooses to create their own, they have an example. If we wanted the notebook to handle more complex cases, I’d have to think of the best way to do so and make edits. 

If there is no issues/concerns with the changes made above, I think the notebook should be good to go.

rizeladiaz commented:
@sosey and @drlaw1558  I will let you both be the last reviewers of this Notebook. I am OK with the comments/changes from Kayli

drlaw1558 commented:
Hi @kglidic - this is looking good to me!  A few additional minor requests and comments:
- Please update to 1.17.1 instead of 1.16.0.  This will both solve an outstanding gwcs issue with 1.15.1 and 1.16.1 and also ensure that we're caught up to the latest pipeline.
- It looks like you've moved the notebook into the NIRSPEC/FSlit/ folder so please delete the old NIRSPEC/Slit/ folder
- Please remove the (commented-out) option to set ramp_fit to use multiple cores.  This should no longer be necessary and we're trying to deprecate it.
- Can you please add some (commented-out is ok) code showing how to enable the clean_flicker_noise step in detector1?
- Can you please move the spec2_asn_manual and spec3_manual (should be spec3_asn_manual for consistency?) to the top cell below where master_bg and pixel_bg are configured?  These are sufficiently key processing directives related to the available files that it would be helpful to be able to set them without digging into the later notebook cells.

- I'm also running into some errors with the downloading of demo files.  I think the issue is that the Observations.download_file call is just giving the directory path but needs to give the file name to write as well.  I believe you need (e.g.)
sci_manifest = Observations.download_file(file, local_path=os.path.join(uncal_dir, Path(file).name))
where you previous will need to
from pathlib import Path

- Looks like the calls to write_asn in spec2 and spec3 needs to use 'bg_files' and not 'background_files'

- If I fix the above, I then get a crash in spec2 
stpipe.Spec2Pipeline.assign_wcs - CRITICAL - No open slits fall on detector NRS2.
Should the manual ASN generator not be creating asn files for NRS2 given that all of the (primary) science slits are on NRS1?

kglidic commented:
Hi @drlaw1558 working on these edit now will push the edits later today or tomorrow.

kglidic commented:
In addition to the demo dataset PID 1128 Obs006, I ran the notebook with different setups from Obs004, and Obs011 of the same program. I addressed the following issues in this update:

- Notebook was updated to pipeline version 1.17.1.
- Removed the commented out line for multiprocessing with ramp_fit in stage 1.
- Added some commented out lines for how to run clean_flicker_noise in stage 1.
- Added try-except to stage 2 processing cells so it does not crash when no slits fall on NRS2.
- Removed old helper functions that are no longer needed since we create the ASN files within the notebook. Added new helper functions to do so.
-  Added in cells that create spec2/spec3 ASN files and changed the section description to match. These cells also include filtering needed for some FS datasets when creating the ASN files (spectral/spatial dithers, ALLSLITS, etc.)
- The ASN file directory and any mentions of it in the notebook were removed as we now create spec2 and spec3 ASN files within the notebook.
- In Stage 2/3 I made subsections that break up the science and background processing. 
- Fixed mis-numbered time benchmarks.

drlaw1558 reviewed (approved):
LG2M!  Merging.
